saml/lib/Saml.php

```diff
@@ -135,6 +135,59 @@ function processLoginResponse($request)
             );
         }
 
+        if ($this->config->isUserGroupsSettingEnabled()) {
+            $samlUserGroupsSetting = $this->config->getUserGroupsSetting();
+            if (isset($samlUserGroupsSetting['attribute']) && $samlUserGroupsSetting['attribute'] != ''
+                && isset($samlUserGroupsSetting['separator']) && $samlUserGroupsSetting['separator'] != ''
+                && isset($attributes[$samlUserGroupsSetting['attribute']])) {
+
+                $allGroups = array();
+                foreach(jAcl2DbUserGroup::getGroupList() as $group) {
+                    if ($group->grouptype == 2) {
+                        // private group, let's ignore
+                        continue;
+                    }
+
+                    $allGroups[] = $group->id_aclgrp;
+                }
+
+                $groupsOfUser = array_map('trim', explode($samlUserGroupsSetting['separator'], $attributes[$samlUserGroupsSetting['attribute']]));
+                if (isset($samlUserGroupsSetting['prefix']) && $samlUserGroupsSetting['prefix'] != '') {
+                    $prefix = $samlUserGroupsSetting['prefix'];
+                    $groupsOfUser = array_map(
+                        function($g) use ($prefix) {
+                            return substr($g, strlen($prefix));
+                        },
+                        array_filter($groupsOfUser, function($g) use ($prefix) {
+                            return substr($g, 0, strlen($prefix)) == $prefix;
+                        })
+                    );
+                }
+
+                $groupsToRemove = array_filter($allGroups, function($g) use ($groupsOfUser) {
+                    return !array_search($g, $groupsOfUser);
+                });
+                $groupsToAdd = array_filter($groupsOfUser,function($g) use ($allGroups) {
+                    return !array_search($g, $allGroups);
```

Comment:
this test is wrong I guess. It should be the contrary, `return array_search($g, $allGroups);`. Right?

The goal is to have a list of groups that **exist**, right?

Reply:
I have updated this filter and used login groups.

I add comment to explicite filters.